_posts/2022-03-21-getting-started-with-protobufs-for-typescript.md

```diff
@@ -0,0 +1,231 @@
+  - On my Mac I downloaded `protoc-x.xx.x-osx-x86_64.zip`, unzipped and copied the `protoc` file in to my `$PATH` at `/usr/local/bin/`.
+
+However, this only generates code for [some languages](https://github.com/protocolbuffers/protobuf#protobuf-runtime-installation). From that list we're interested in JavaScript. The bit that isn’t included with `protoc` is generating the Typescript types. For that you need a plugin for `protoc` like the [`ts-proto`](https://www.npmjs.com/package/ts-proto) npm package.
+
+Let's start building.  Start a TypeScript project and create an `addressBook.proto` file as above in the same project. Then run:
+
+- `npm install ts-protoc-gen`
+
+Next compile your `.proto` file into a `.ts` file for use within your project by running the following CLI command:
+
+```
+protoc \
+    --plugin="./node_modules/.bin/protoc-gen-ts_proto" \
+    --ts_proto_opt=esModuleInterop=true \
+    --ts_proto_out="./src/generated" \
+    src/proto/addressBook.proto
+```
+
+This is the point where I really like TypeScript. If you run `protoc` without the `ts-proto` plugin and look at the generated code there are 760 lines of stuff in the `generated/src/proto/addressBook.js` file which honestly is not that easy to read or understand how to use. But if you look at the `generated/src/proto/addressBook.ts` file, there are only 250 lines and it provides you with the type interfaces and commands you are going to need to create and use your messages and Protobufs.
+
+_*This is where I fell over on my first attempt, I did not use a plugin like `ts-proto` to generate the types and only had the JS output. I couldn't easily enough understand what was going on, got lost and then distracted by other shiny stuff.*_
+
+## Create a Message
+_*When you read the docs and get presented with ["The API is not well-documented yet"](https://github.com/protocolbuffers/protobuf/tree/master/js#api) you can be forgiven for wanting to cry and giving up. This is where I failed on my second attempt and this is where the Typescript types come to the rescue.*_
+
+Creating the contents of your message is now as easy as:
+```js
+import { AddressBook, Person_PhoneType } from './generated/src/proto/addressBook'
+
+const myAddressBook: AddressBook = {
+  people: [
+    {
+      name: "Joe Blogs",
+      phones: [
+        {
+          phoneNumber: "0123456789",
+          phoneType: Person_PhoneType.MOBILE,
+        }
+      ]
+    },
+    {
+      name: "Jane Smith",
+      phones: [
+        {
+          phoneNumber: "0987654321",
+          phoneType: Person_PhoneType.HOME,
+        }
+      ]
+    }
+  ]
+}
+```
+This is almost identical to the original JSON object used in the [data structure section above](#define-the-data-structure-for-your-messages), how fantastic is that!  There are two differences in the above that take advantage of some awesome idiomatic Typescript and make life really easy:
```

Comment:
```suggestion
This is almost identical to the original JSON object used in the [data structure section above](#define-the-data-structure-for-your-messages), how fantastic is that?  There are two differences in the above that take advantage of some awesome idiomatic Typescript and make life really easy:
```

nitpick: You're asking the reader a question

If it's a shouty question, then ?! works for me, but others may find it jarring (personal style).

I have no issues with !!!!! (yes, count them) exclamation marks.  Bonus points if you get the reference ;)

Reply:
Hmmm.  This is intended as a rhetorical question so !/./? are all acceptable.  This was included for emphasis hence '!' but as I am trying to challenge the reader to also think about the comparison I will change to '?'. 

"Five exclamation marks, the sure sign of an insane mind."